x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/stats/Metrics.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.sql.stats;
+
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.xpack.core.watcher.common.stats.Counters;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Class encapsulating the metrics collected for ES SQL
+ */
+public class Metrics {
+    private enum OperationType {
+        FAILED, PAGING, TOTAL;
+
+        @Override
+        public String toString() {
+            return this.name().toLowerCase(Locale.ROOT);
+        }
+    }
+    
+    // map that holds total/paging/failed counters for each client type (rest, cli, jdbc, odbc...)
+    private final Map<QueryMetric, Map<OperationType, CounterMetric>> opsByTypeMetrics = new HashMap<>();
+    // map that holds one counter per sql query "feature" (having, limit, order by, group by...) 
+    private final Map<FeatureMetric, CounterMetric> featuresMetrics = new HashMap<>();
+    // because the hook in the Verifier/Analyzer is being called several times during the query analysis flow
+    // we keep a set of flags so that we don't count a certain metric more than once
+    private final Map<FeatureMetric, Boolean> featuresCollected = new HashMap<>();
```

Comment:
You could use a HashSet and check existence with `contains()`

Reply:
Just use  `map.putIfAbsent()`